### PR TITLE
Ensure we return non-negative scores when scoring scalar dot-products

### LINKED ISSUE #108339: [CI] CcsCommonYamlTestSuiteIT test {p0=search.vectors/41_knn_search_byte_quantized/Test create, merge, and search dot_product} failing
This fails on release-tests if that makes any difference.

The failure is probably not specifically about this test since it has a bunch other tests failed with the same test cluster and it has the following AssertionError

```
[2024-05-07T16:51:03,414][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [remote_cluster-1] fatal error in thread [elasticsearch[remote_cluster-1][search_worker][T#24]], exiting java.lang.AssertionError	
	at org.apache.lucene.core@9.10.0/org.apache.lucene.search.TopScoreDocCollector$SimpleTopScoreDocCollector$1.collect(TopScoreDocCollector.java:75)	
	at org.apache.lucene.core@9.10.0/org.apache.lucene.search.Weight$DefaultBulkScorer.scoreRange(Weight.java:277)	
	at org.apache.lucene.core@9.10.0/org.apache.lucene.search.Weight$DefaultBulkScorer.score(Weight.java:236)	
	at org.elasticsearch.server@8.14.0/org.elasticsearch.search.internal.CancellableBulkScorer.score(CancellableBulkScorer.java:45)	
	at org.apache.lucene.core@9.10.0/org.apache.lucene.search.BulkScorer.score(BulkScorer.java:38)	
	at org.elasticsearch.server@8.14.0/org.elasticsearch.search.internal.ContextIndexSearcher.searchLeaf(ContextIndexSearcher.java:431)	
	at org.elasticsearch.server@8.14.0/org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:360)	
	at org.elasticsearch.server@8.14.0/org.elasticsearch.search.internal.ContextIndexSearcher.lambda$search$4(ContextIndexSearcher.java:345)	
	at org.apache.lucene.core@9.10.0/org.apache.lucene.search.TaskExecutor$TaskGroup.lambda$createTask$0(TaskExecutor.java:117)	
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
```

**Build scan:**
https://gradle-enterprise.elastic.co/s/vgcc7jbcvzozo/tests/:qa:ccs-common-rest:yamlRestTest/org.elasticsearch.test.rest.yaml.CcsCommonYamlTestSuiteIT/test%20%7Bp0=search.vectors%2F41_knn_search_byte_quantized%2FTest%20create,%20merge,%20and%20search%20dot_product%7D

**Reproduction line:**
```
./gradlew ':qa:ccs-common-rest:yamlRestTest' --tests "org.elasticsearch.test.rest.yaml.CcsCommonYamlTestSuiteIT" -Dtests.method="test {p0=search.vectors/41_knn_search_byte_quantized/Test create, merge, and search dot_product}" -Dtests.seed=CA15C0087EBCFB79 -Dbuild.snapshot=false -Dtests.jvm.argline="-Dbuild.snapshot=false" -Dtests.locale=is-IS -Dtests.timezone=Pacific/Kwajalein -Druntime.java=21
```

**Applicable branches:**
8.14

**Reproduces locally?:**
Didn't try

**Failure history:**
[Failure dashboard for `org.elasticsearch.test.rest.yaml.CcsCommonYamlTestSuiteIT#test {p0=search.vectors/41_knn_search_byte_quantized/Test create, merge, and search dot_product}`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('test%20%7Bp0%3Dsearch.vectors/41_knn_search_byte_quantized/Test%20create,%20merge,%20and%20search%20dot_product%7D'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.test.rest.yaml.CcsCommonYamlTestSuiteIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure excerpt:**
```
java.net.ConnectException: Connection refused

  at __randomizedtesting.SeedInfo.seed([CA15C0087EBCFB79:4241FFD2D0409681]:0)
  at org.elasticsearch.client.RestClient.extractAndWrapCause(RestClient.java:934)
  at org.elasticsearch.client.RestClient.performRequest(RestClient.java:304)
  at org.elasticsearch.client.RestClient.performRequest(RestClient.java:307)
  at org.elasticsearch.client.RestClient.performRequest(RestClient.java:292)
  at org.elasticsearch.test.rest.ESRestTestCase.ensureNoInitializingShards(ESRestTestCase.java:1734)
  at org.elasticsearch.test.rest.ESRestTestCase.cleanUpCluster(ESRestTestCase.java:530)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:1004)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.elasticsearch.test.cluster.local.DefaultLocalElasticsearchCluster$1.evaluate(DefaultLocalElasticsearchCluster.java:47)
  at org.elasticsearch.test.cluster.local.DefaultLocalElasticsearchCluster$1.evaluate(DefaultLocalElasticsearchCluster.java:47)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

  Caused by: java.net.ConnectException: Connection refused

    at sun.nio.ch.Net.pollConnect(Net.java:-2)
    at sun.nio.ch.Net.pollConnectNow(Net.java:682)
    at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:973)
    at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processEvent(DefaultConnectingIOReactor.java:174)
    at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processEvents(DefaultConnectingIOReactor.java:148)
    at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor.execute(AbstractMultiworkerIOReactor.java:351)
    at org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager.execute(PoolingNHttpClientConnectionManager.java:221)
    at org.apache.http.impl.nio.client.CloseableHttpAsyncClientBase$1.run(CloseableHttpAsyncClientBase.java:64)
    at java.lang.Thread.run(Thread.java:1583)

```

### PR DESCRIPTION
**Pull request by benwtrent**
closes: https://github.com/elastic/elasticsearch/issues/108339

### CODE REVIEW

**Review by carlosdelest — APPROVED** [review]

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
The scorer should be doing the same as that of the lucene one. Why is this necessary? Also, should the lucene one do same (in lucene itself)?

**Comment by ChrisHegarty** [comment]
Since the individual vector values are positive, then is the negativity coming from the offsets or correction values? I think our tests should be updated to cover this.

**Comment by benwtrent** [comment]
>  then is the negativity coming from the offsets or correction values? I think our tests should be updated to cover this.

This is being fixed because of a failing test. I could add a manual one where vectors are `0` & offset value is negative.

I also have a PR open in Lucene to correct this there. But, here, we have our own scorer implementation that needs handling.

One thing I should do here is do an `assert >=0` on the scalar dot product to ensure we aren't allowing some sneaky bugs throught.

**Comment by benwtrent** [comment]
@ChrisHegarty added a test & assertion that our dot product from the native impl is `>=0`

File: libs/vec/src/test/java/org/elasticsearch/vec/VectorScorerFactoryTests.java [hunk 1]
```diff
@@ -92,6 +93,51 @@ void testSimpleImpl(long maxChunkSize) throws IOException {
...
+                assertThat(scorer.score(0, 1), greaterThanOrEqualTo(0f));
+                assertThat((new VectorScorerSupplierAdapter(scorer)).scorer(0).score(1), equalTo(expected));
+                // max inner product
+                expected = luceneScore(MAXIMUM_INNER_PRODUCT, vec1, vec2, 1, -5, -5);
+                scorer = factory.getInt7ScalarQuantizedVectorScorer(32, 2, 1, MAXIMUM_INNER_PRODUCT, in).get();
+                assertThat(scorer.score(0, 1), greaterThanOrEqualTo(0f));
+                assertThat(scorer.score(0, 1), equalTo(expected));
+                assertThat((new VectorScorerSupplierAdapter(scorer)).scorer(0).score(1), equalTo(expected));
+                // cosine
+                expected = 0f; // TODO fix in Lucene: https://github.com/apache/lucene/pull/13356 luceneScore(COSINE, vec1, vec2, 1, -5,
```
Comment (ChrisHegarty) [inline libs/vec/src/test/java/org/elasticsearch/vec/VectorScorerFactoryTests.java:129]: ah ha. I see this now. Thanks

**Comment by ChrisHegarty** [comment]
>  I could add a manual one where vectors are `0` & offset value is negative.

Thanks, this is helpful.

> I also have a PR open in Lucene to correct this there. But, here, we have our own scorer implementation that needs handling.

Ok cool. The point is that our scorer implementation should produce the same result as that of the Lucene implementation. I now see the separate Lucene PR, thanks.

> One thing I should do here is do an `assert >=0` on the scalar dot product to ensure we aren't allowing some sneaky bugs throught.

Yeah. Since asserts bloat the bytecode, an alternative is to put the assert in the tests that are closer to the implementation, so in `JDKVectorLibraryTests`.

**Review by ChrisHegarty — APPROVED** [review]

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Pull request merged**